File: packages/server/src/agent/SqlAgent.ts
/* The provided code snippet is defining a TypeScript class called `SqlAgent` that serves as a service in a NestJS application. This class is responsible for interacting with an AI agent to generate SQL
queries based on user input. */
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { llm, type BaseLLM } from './llm.js';

import { getSqlPrompt, getChartPrompt } from './prompts/index.js';
import { DatabaseService } from '../modules/database/database.service.js';
import {
  CotAgent,
  FieldType,
  InputField,
  OutputField,
  Signature,
} from './CotAgent.js';

@Injectable()
export class SqlAgent {
  private readonly logger = new Logger(SqlAgent.name);

  private readonly lastMessageIds = new Map<string, string>();

  llm: BaseLLM;
  sig: Signature;
  cot: CotAgent;

  constructor(
    private config: ConfigService,
    private databaseService: DatabaseService,
  ) {
    if (this.config.get('OPENAI_API_KEY')) {
      this.llm = llm;
    } else {
      throw new Error('No API key provided');
    }
    this.llm.init();

    const sig = new Signature({
      messages: new InputField({
        type: FieldType.ListDict,
        description: 'List of messages in the conversation.',
      }),
      reasoning: new OutputField({
        type: FieldType.Str,
        description: '',
      }),
      answer: new OutputField({
        type: FieldType.Str,
        description: 'The SQL query, do not use markdown.',
      }),
      reason: new OutputField({ type: FieldType.Str, description: '' }),
    });

    this.cot = new CotAgent(sig, this.llm);
  }

  $onInit() {
    this.logger.debug('Init SqlAgent');
  }

  public async run(question: string, conversationId: string) {
    // 1. Get all table schemas for generate context prompt
    const tableSchemas = await this.databaseService.getAllTableSchemas();
    const dbType = await this.databaseService.getDbType();

    // 2. Generate SQL prompt
    const messages = getSqlPrompt(
      question,
      JSON.stringify(tableSchemas),
      dbType,
    );

    // 3. Send question prompt to AI agent
    const {
      object,
      response: { id },
    } = await this.cot.run(messages);

    // cache last message id
    // this.lastMessageIds.set(conversationId, contextResponse.id);

    // this.lastMessageIds.set(conversationId, msgId);

    return {
      id,
      type: 'sql',
      ...object,
    };
  }
}

File: chatbi/agent/sql_agent.py
import dspy
from loguru import logger

from chatbi.agent.agent_message import (
    AgentMessage,
)
from chatbi.agent.llm import LLM
from chatbi.agent.prompts.sql_prompt import get_sql_prompt

dspy.enable_logging()


class SqlSignature(dspy.Signature):
    """Signature for the text2sql method."""

    context = dspy.InputField(desc="The context of the question, e.g. table schema.")
    question = dspy.InputField()
    answer = dspy.OutputField(desc="The PostgreSQL SQL query, do not use markdown.")
    reason = dspy.OutputField()


class SqlSignatureV2(dspy.Signature):
    """Signature for the text2sql method."""

    messages: list[dict] = dspy.InputField()
    answer = dspy.OutputField(desc="The PostgreSQL SQL query, do not use markdown.")
    reason = dspy.OutputField()


class SqlAgent:
    def __init__(self, name="SqlAgent"):
        self.name = name

        # initialize the LLM model
        LLM["openai"].get_model()

    def reply(self, id: str, question: str, table_schema) -> AgentMessage:
        logger.debug(f"[{self.name}]: infer sql")

        infer = dspy.ChainOfThought(SqlSignatureV2)
        messages = get_sql_prompt(question=question, table_schema=table_schema)

        logger.debug(f"[{self.name}]: messages: {messages}")
        # response = self.model.llm(messages=messages)
        response = infer(messages=messages)

        answer = response["answer"]
        reason = response["reason"]
        reasoning = response["reasoning"]

        logger.debug(f"[{self.name}]: response: {response}")

        return AgentMessage(
            name=self.name,
            id=id,
            type="sql",
            answer=answer,
            reason=reason,
            reasoning=reasoning,
        )
